x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/NodeTermEnumRequest.java

```diff
@@ -0,0 +1,160 @@
+
+import org.elasticsearch.action.IndicesRequest;
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.transport.TransportRequest;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Internal termenum request executed directly against a specific node, querying potentially many 
+ * shards in one request
+ */
+public class NodeTermEnumRequest extends TransportRequest implements IndicesRequest {
+
+    private String field;
+    private String string;
+    private long taskStartedTimeMillis;
+    private long nodeStartedTimeMillis;
+    private boolean caseInsensitive;
+    private int size;
+    private long timeout;
+    private final QueryBuilder indexFilter;
+    private Set<ShardId> shardIds;
+    private String nodeId;
+    
+
+    public NodeTermEnumRequest(StreamInput in) throws IOException {
+        super(in);
+        // Set the clock running as soon as we appear on a node.
+        nodeStartedTimeMillis = System.currentTimeMillis();
+
+        field = in.readString();
+        string = in.readString();
+        caseInsensitive = in.readBoolean();
+        size = in.readVInt();
+        timeout = in.readVLong();
+        taskStartedTimeMillis = in.readVLong();
+        indexFilter = in.readOptionalNamedWriteable(QueryBuilder.class);
+        nodeId = in.readString();
+        int numShards = in.readVInt();
+        shardIds = new HashSet<>(numShards);
+        for (int i = 0; i < numShards; i++) {
+            shardIds.add(new ShardId(in));
+        }
+    }
+
+    public NodeTermEnumRequest(final String nodeId, final Set<ShardId> shardIds, TermEnumRequest request) {
+        this.field = request.field();
+        this.string = request.string();
+        this.caseInsensitive = request.caseInsensitive();
+        this.size = request.size();
+        this.timeout = request.timeout().getMillis();
+        this.taskStartedTimeMillis = request.taskStartTimeMillis;
+        this.indexFilter = request.indexFilter();
+        this.nodeId = nodeId;
+        this.shardIds = shardIds;        
+        
+        // TODO serialize shard ids
+    }
+
+    public String field() {
+        return field;
+    }
+
+    public String string() {
+        return string;
+    }
+
+    public long taskStartedTimeMillis() {
+        return this.taskStartedTimeMillis;
+    }
+    
+    /** 
+     * The time this request was materialized on a shard
+     * (defaults to "now" if serialization was not used e.g. a local request).
+     */
+    public long shardStartedTimeMillis() {
+        if (nodeStartedTimeMillis == 0) {
+            nodeStartedTimeMillis = System.currentTimeMillis();
+        }
+        return this.nodeStartedTimeMillis;
+    }    
```

Comment:
I'm confused why the method is called `shardStartedXXX` while the field is called `nodeStartedXXX`.

Reply:
A relic from when we did per-shard requests rather than bundling into a per-node request